test/test_sync.py

```diff
@@ -0,0 +1,305 @@
+    assert_expected_archive_paths(sync.source_paths,
+                                  expected_outputs,
+                                  expected_protected_outputs)
+
+    # Tidy up test - Remove any added enviroment variables
+    for envt_var in envt_vars.keys():
+        del os.environ[envt_var]
+
+
+@pytest.mark.parametrize(
+    "add_config, envt_vars, expected_restarts, expected_protected_restarts",
+    [
+        (
+            {
+                "sync": {
+                    'restarts': True
+                },
+                "restart_freq": 5
+            }, {},
+            [], ['restart000']
+        ),
+        (
+            {
+                "sync": {
+                    'restarts': True
+                },
+                "restart_freq": 2
+            }, {},
+            ['restart000', 'restart002'], ['restart004']
+        ),
+        (
+            {
+                "sync": {
+                    "restarts": True
+                },
+                "restart_freq": 2
+            }, {'PAYU_SYNC_IGNORE_LAST': 'True'},
+            ['restart000', 'restart002'], []
+        ),
+        (
+            {"restart_freq": 3}, {'PAYU_SYNC_RESTARTS': 'True'},
+            ['restart000', 'restart001', 'restart002'],
+            ['restart003', 'restart004']
+        ),
+    ])
+def test_restarts_to_sync(add_config, envt_vars,
+                          expected_restarts, expected_protected_restarts):
+    sync = setup_sync(add_config, envt_vars)
+
+    # Test function
+    sync.add_restarts_to_sync()
+
+    # Assert expected restarts and protected restarts are added
+    assert_expected_archive_paths(sync.source_paths,
+                                  expected_restarts,
+                                  expected_protected_restarts)
+
+    # Tidy up test - Remove any added enviroment variables
+    for envt_var in envt_vars.keys():
+        del os.environ[envt_var]
+
+
+def test_rsync_components():
+    additional_config = {
+        "sync": {
+            "rsync_flags": "--compress",
+            "rsync_protocol": 29,
+            "url": "test.domain",
+            "user": "test-usr",
+            "path": "remote/path",
+            "exclude": ["iceh.????-??-??.nc", "*-DEPRECATED"]
+        }}
+    sync = setup_sync(additional_config=additional_config)
+
+    # Test base_rsync_cmd
+    sync.set_base_rsync_cmd()
+
+    home_path = os.getenv('HOME')
+    expected_cmd = ('rsync -vrltoD --safe-links --compress --protocol=29'
+                    f' -e "ssh -i {home_path}/.ssh/id_rsa_file_transfer"')
+
+    assert sync.base_rsync_cmd == expected_cmd
+
+    # Test destination_path
+    sync.set_destination_path()
+    assert sync.destination_path == "test-usr@test.domain:remote/path"
+
+    # Test excludes
+    sync.set_excludes_flags()
+    expected_excludes = "--exclude iceh.????-??-??.nc --exclude *-DEPRECATED"
+    expected_excludes += " --exclude *.nc.*"
+    assert sync.excludes == expected_excludes
```

Comment:
Using a class as you have makes these tests really neat and clean. Very readable.